lib/client/keystore.go

```diff
@@ -203,493 +222,355 @@ func (fs *FSLocalKeyStore) AddKey(host, username string, key *Key) error {
-		if ok {
-			stored.KubeTLSCerts = nil
-		}
-	default:
-		return trace.BadParameter("unexpected key store type %T", store)
 	}
 	return nil
 }
 
-// WithDBCerts returns a GetKeyOption to load database access certificates
-// from the store for a given Teleport cluster.
-func WithDBCerts(teleportClusterName, dbName string) KeyOption {
-	return withDBCerts{teleportClusterName: teleportClusterName, dbName: dbName}
+func (o WithSSHCerts) deleteFromKey(key *Key) {
+	key.Cert = nil
 }
 
-type withDBCerts struct {
-	teleportClusterName, dbName string
-}
+// WithKubeCerts is a CertOption for handling kubernetes certificates.
+type WithKubeCerts struct{}
 
-func (o withDBCerts) getKey(store LocalKeyStore, idx keyIndex, key *Key) error {
-	switch s := store.(type) {
-	case *FSLocalKeyStore:
-		dirPath := s.dirFor(idx.proxyHost)
-		dbDir := filepath.Join(dirPath, idx.username+dbDirSuffix, o.teleportClusterName)
-		dbFiles, err := ioutil.ReadDir(dbDir)
-		if err != nil && !os.IsNotExist(err) {
-			return trace.Wrap(err)
-		}
-		if key.DBTLSCerts == nil {
-			key.DBTLSCerts = make(map[string][]byte)
-		}
-		for _, fi := range dbFiles {
-			data, err := ioutil.ReadFile(filepath.Join(dbDir, fi.Name()))
-			if err != nil {
-				return trace.Wrap(err)
-			}
-			dbName := strings.TrimSuffix(filepath.Base(fi.Name()), fileExtTLSCert)
-			key.DBTLSCerts[dbName] = data
-		}
-	case *MemLocalKeyStore:
-		idx.clusterName = o.teleportClusterName
-		stored, ok := s.inMem[idx]
-		if !ok {
-			return trace.NotFound("key for %v not found", idx)
-		}
-		key.DBTLSCerts = stored.DBTLSCerts
-	default:
-		return trace.BadParameter("unexpected key store type %T", store)
+func (o WithKubeCerts) relativeCertPath(idx KeyIndex) string {
+	components := []string{idx.ProxyHost, idx.Username + kubeDirSuffix}
+	if idx.ClusterName != "" {
+		components = append(components, idx.ClusterName)
 	}
-	if key.ClusterName == "" {
-		key.ClusterName = o.teleportClusterName
-	}
-	return nil
+	return filepath.Join(components...)
 }
 
-func (o withDBCerts) deleteKey(store LocalKeyStore, idx keyIndex) error {
-	// If database name is specified, remove only that cert, otherwise remove
-	// certs for all databases a user is logged into.
-	switch s := store.(type) {
-	case *FSLocalKeyStore:
-		dirPath := s.dirFor(idx.proxyHost)
-		if o.dbName != "" {
-			return os.Remove(filepath.Join(dirPath, idx.username+dbDirSuffix, o.teleportClusterName, o.dbName+fileExtTLSCert))
-		}
-		return os.RemoveAll(filepath.Join(dirPath, idx.username+dbDirSuffix, o.teleportClusterName))
-	case *MemLocalKeyStore:
-		idx.clusterName = o.teleportClusterName
-		stored, ok := s.inMem[idx]
-		if !ok {
-			return trace.NotFound("key for %v not found", idx)
-		}
-		if o.dbName != "" {
-			stored.DBTLSCerts[o.dbName] = nil
-		} else {
-			stored.DBTLSCerts = nil
-		}
-	default:
-		return trace.BadParameter("unexpected key store type %T", store)
+func (o WithKubeCerts) updateKey(key *Key, certData interface{}) error {
```

Comment:
Do not pass interface{} please. Accept exact type.

Reply:
What do you propose that exact type to be? As I see it, since both `[]byte` and `map[string][]byte` need to be accepted with this uniform `updateKey` interface, there is no other way around it except to bring back the non-trivial amount of boilerplate duplication as it used to be the case before.

Reply:
@andrejtokarcik I can see two options:

* Have two separate methods, `updateKeyByte(*Key, []byte)` and `updateKeyMap(*Key, map[string][]byte)`. Not ideal probably and will result in some code duplication but better than `interface{}`.
* Have this method accept a struct like
```go
type KeyData struct { // Or whatever makes sense to call it.
  ByteData []byte 
  MapData map[string][]byte // <-- I'd rename these field to make more sense.
}
```
and explicitly check which one is set in this method.